net_transport_test.go

```diff
@@ -467,6 +439,105 @@ func TestNetworkTransport_AppendEntriesPipeline_CloseStreams(t *testing.T) {
 	}
 }
 
+func TestNetworkTransport_AppendEntriesPipeline_MaxRPCsInFlight(t *testing.T) {
+	// Test the important cases 0 (default to 2), 1 (disabled), 2 and "some"
+	for _, max := range []int{0, 1, 2, 10} {
+		t.Run(fmt.Sprintf("max=%d", max), func(t *testing.T) {
+			config := &NetworkTransportConfig{
+				MaxPool:         2,
+				MaxRPCsInFlight: max,
+				Timeout:         time.Second,
+				// Don't use test logger as the transport has multiple goroutines and
+				// causes panics.
+				ServerAddressProvider: &testAddrProvider{"localhost:0"},
+			}
+
+			// Transport 1 is consumer
+			trans1, err := NewTCPTransportWithConfig("localhost:0", nil, config)
+			require.NoError(t, err)
+			defer trans1.Close()
+
+			// Make the RPC request
+			args := makeAppendRPC()
+			resp := makeAppendRPCResponse()
+
+			ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
+			defer cancel()
+
+			// Transport 2 makes outbound request
+			config.ServerAddressProvider = &testAddrProvider{string(trans1.LocalAddr())}
+			trans2, err := NewTCPTransportWithConfig("localhost:0", nil, config)
+			require.NoError(t, err)
+			defer trans2.Close()
+
+			// Kill the transports on the timeout to unblock. That means things that
+			// shouldn't have blocked did block.
+			go func() {
+				<-ctx.Done()
+				trans2.Close()
+				trans1.Close()
+			}()
+
+			// Attempt to pipeline
+			pipeline, err := trans2.AppendEntriesPipeline("id1", trans1.LocalAddr())
+			if max == 1 {
+				// Max == 1 implies no pipelining
+				require.EqualError(t, err, ErrPipelineReplicationNotSupported.Error())
+				return
+			}
+			require.NoError(t, err)
+
+			expectedMax := max
+			if max == 0 {
+				// Should have defaulted to 2
+				expectedMax = 2
+			}
+
+			for i := 0; i < expectedMax-1; i++ {
+				// We should be able to send `max - 1` rpcs before `AppendEntries`
+				// blocks. It blocks on the `max` one because it it sends before pushing
+				// to the chan. It will block forever when it does because nothing is
+				// responding yet.
+				out := new(AppendEntriesResponse)
+				_, err := pipeline.AppendEntries(&args, out)
+				require.NoError(t, err)
+			}
+
+			// Verify the next send blocks without blocking test forever
+			errCh := make(chan error, 1)
+			go func() {
+				out := new(AppendEntriesResponse)
+				_, err := pipeline.AppendEntries(&args, out)
+				errCh <- err
+			}()
+
+			select {
+			case <-errCh:
```

Comment:
There is a bug here := we aren't capturing the error from the channel to use in the assertion below...